service-userlayer/src/main/java/org/oskari/map/userlayer/input/KMLParser.java

```diff
@@ -28,24 +40,126 @@
+            // KML always lon,lat 4326
+            CoordinateReferenceSystem sourceCRS = CRS.decode("EPSG:4326", true);
+            MathTransform transform = FeatureCollectionParsers.getTransform(sourceCRS, targetCRS);
+
+            if (extendedData.isEmpty()){
+               builder = getBasicBuilder(targetCRS);
+            } else {
+               builder = getExtendedBuilder(targetCRS, extendedData);
+            }
+            SimpleFeatureIterator iter = fc_kml.features();
+            while (iter.hasNext()){
+                f = iter.next();
+                builder.set(KML_NAME, f.getAttribute("name"));
+                // Basic builder
+                if (extendedData.isEmpty()){
+                    builder.set(KML_DESC, f.getAttribute("description"));
+                // Extended builder
+                } else {
+                    Map<String, Object> untypedData = (Map<String, Object>) f.getUserData().get("UntypedExtendedData");
+                    if (untypedData != null){
+                        extendedData.forEach((k) -> {
+                            if (k.equals("description")){
+                                builder.set(KML_DESC, untypedData.get(k));
+                                return;
+                            }
+                            builder.set(k, untypedData.get(k));
+                        });
+                    } else if (f.getAttribute("description") != null) {
+                        builder.set(KML_DESC, f.getAttribute("description"));
+                    }
+                }
+                Geometry geom = (Geometry) f.getDefaultGeometry();
+                if (geom != null) {
+                    builder.set(KML_GEOM, JTS.transform(geom, transform));
+                }
+                fc.add(builder.buildFeature(f.getID()));
+            }
+            return fc;
+        } catch (Exception e) {
+            throw new ServiceException("Failed to parse KML", e);
+        }
+    }
+    private SimpleFeatureBuilder getBasicBuilder (CoordinateReferenceSystem targetCRS) {
+        SimpleFeatureTypeBuilder builder = new SimpleFeatureTypeBuilder();
+        builder.setName( "BasciKMLBuilder" );
+        builder.setNamespaceURI( "http://www.oskari.org" );
+        builder.setCRS(targetCRS);
+        builder.add(KML_NAME, String.class );
+        builder.add(KML_DESC, String.class );
+        builder.add(KML_GEOM, Geometry.class );
+        return new SimpleFeatureBuilder(builder.buildFeatureType());
+    }
+    private SimpleFeatureBuilder getExtendedBuilder (CoordinateReferenceSystem targetCRS, Set<String> extendedData) {
+        SimpleFeatureTypeBuilder builder = new SimpleFeatureTypeBuilder();
+        builder.setName( "ExtendedKMLBuilder" );
+        builder.setNamespaceURI( "http://www.oskari.org" );
+        builder.setCRS(targetCRS);
+        builder.add(KML_NAME, String.class );
+        builder.add(KML_DESC, String.class );
+        extendedData.forEach((k) -> {
```

Comment:
Can these be combined to a single method with just checking if extendedData is null -> don't do the loop? Does the builder name matter? Looks like the method's are the same other than that.